controllers/prometheus_exporter.go

```diff
@@ -0,0 +1,202 @@
+    label["app"] = mdb.Name() + '-exporter'
+
+    password := ''
+    for _, user := range mdb.Users {
+        if user.Name() == exporterUserName {
+            secretsClient := r.client.CoreV1().Secrets(mdb.Namespace)
+            passwordRef := corev1.Secret{}
+            err := secretsClient.Get(context.TODO(), user.PasswordSecretRef.Name(), %passwordRef)
+            password = passwordRef.Data[user.PasswordSecretRef.Key]
+            break
+        }
+    }
+
+    full_uri = fmt.Sprintf(
+        "mongodb://%s:%s@%s.%s.%s.%s:%d",
+        exporterUserName,
+        password,
+        m.ServiceName(),
+        m.Namespace,
+        clusterDomain,
+        27017)
+
+    return secret.Builder().
+        SetName(mdb.Name() + '-exporter-uri').
+        SetNamespace(mdb.Namespace).
+        SetLabels(label).
+        SetField('mongodb-uri', full_uri).
+        Build()
+}
+
+func (r *ReplicaSetReconciler) ensureExporterDeployment(mdb mdbv1.MongoDBCommunity) error {
+    dpl := buildExporterDeployment(mdb)
+    err := r.client.Create(context.TODO(), &dpl)
+    if err != nil && apiErrors.IsAlreadyExists(err) {
+        r.log.Infof("The deployment already exists... moving forward: %s", err)
+        return nil
+    }
+    return err
+}
+
+// buildExporterDeployment creates a Deployment for the Prometheus Exporter
+func buildExporterDeployment(mdb mdbv1.MongoDBCommunity) appsv1.Deployment {
+    label := make(map[string]string)
+    label["app"] = mdb.Name() + '-exporter'
+
+    ExporterServiceAccount := buildExporterServiceAccount(mdb)
+    err := r.client.Create(context.TODO(), &ExporterServiceAccount)
+    if err != nil && !apiErrors.IsAlreadyExists(err) {
+        r.log.Infof("Error when building exporter service account: %s", err)
+        return err
+    }
+
+    MongoDbUriSecret := buildMongoDbUriSecret(mdb)
+    err := r.client.Create(context.TODO(), &MongoDbUriSecret)
+    if err != nil && !apiErrors.IsAlreadyExists(err) {
+        r.log.Infof("Error when building Mongo DB URI Secret: %s", err)
+        return err
+    }
+
+    return deployment.Builder().
+        SetName(mdb.Name() + '-exporter').
+        SetNamespace(mdb.Namespace).
+        SetLabel(label).
+        SetAnnotations().
+        SetReplicas(1).
+        SetSelector(label).
+        SetPodAnnotations().
+        SetPodLabel(label).
+        SetServiceAccount(ExporterServiceAccount.Name()).
+        SetContainerMongoDBURI(MongoDbUriSecret.Name()).
+        SetContainerEnvs().
+        SetContainerImage().
+        SetContainerImagePullPolicy().
+        SetContainerListenerPort().
+        SetContainerExtraArgs().
+        SetContainerLivenessProbe().
+        SetContainerReadinessProbe().
+        SetContainerResources().
+        SetContainerImagePullSecrets().
+        SetContainerNodeSelector().
+        SetContainerTolerations().
+        Build()
+
+func (r *ReplicaSetReconciler) ensureExporterService(mdb mdbv1.MongoDBCommunity) error {
```

Comment:
can you pass context into this? generally in golang you pass a preexisting context around. The point of `context.Context` is it allows you to apply timeouts and cancel calls as necessary.

Reply:
The rest of the code passes `context.TODO()` around which, if I recall correctly, allows golang to determine the context itself "under the hood."